source/s3_buffer_pool.c

```diff
@@ -0,0 +1,225 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/s3/private/s3_buffer_pool.h>
+
+#include <aws/common/array_list.h>
+#include <aws/common/mutex.h>
+
+/*
+ * S3 Buffer Pool.
+ * Fairly trivial implementation of "arena" style allocator.
+ * Note: current implementation is not optimized and instead tries to be
+ * as straightforward as possible. Given that pool manages a small number
+ * of big allocations, performance impact is not that bad, but something we need
+ * to look into on the next iteration.
+ *
+ * Basic approach is to keep a pool of blocks (block size provided by caller,
+ * ex. 128 MB as a typical size). This is referred to as primary storage.
+ * At high level, each block has offset ptr and allocation basically just means
+ * bumping the offset ptr up and increasing blocks alloc count.
+ * Deallocation decreases the block alloc count and if alloc count ever reaches
+ * 0, the offset ptr is reset back to the start of the block.
+ * For acquires less than a block size, the current approach is to go through
+ * all blocks and see if any of them can fit the size and if not a new block is
+ * created.
+ * For acquires over block size, pool falls back to acquiring/releasing through
+ * base allocator.
+ * If memory limit is set, then acquires will return empty buffers when memory
+ * limit is hit. For primary, that means acquires will fail when new block is
+ * needed, but there is no mem for it. For secondary storage, it tracks mem used
+ * by malloc.
+ */
+
+static size_t s_min_num_blocks = 5;
+
+struct aws_s3_buffer_pool {
+    struct aws_allocator *base_allocator;
+    struct aws_mutex mutex;
+
+    size_t block_size;
+    size_t max_mem_usage;
+    size_t current_mem_usage;
+
+    struct aws_array_list blocks;
+};
+
+struct s3_buffer_pool_block {
+    size_t block_size;
+    uint8_t *block_ptr;
+    uint8_t *offset_ptr;
+
+    size_t alloc_count;
+};
+
+struct aws_s3_buffer_pool *aws_s3_buffer_pool_new(
+    struct aws_allocator *allocator,
+    size_t block_size,
+    size_t max_mem_usage) {
+    struct aws_s3_buffer_pool *buffer_pool = aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_buffer_pool));
+
+    AWS_FATAL_ASSERT(buffer_pool != NULL);
+
+    buffer_pool->base_allocator = allocator;
+    buffer_pool->block_size = block_size;
+    buffer_pool->current_mem_usage = 0;
+    buffer_pool->max_mem_usage = max_mem_usage;
+    if (aws_mutex_init(&buffer_pool->mutex)) {
+        goto on_error;
+    }
+
+    aws_array_list_init_dynamic(&buffer_pool->blocks, allocator, s_min_num_blocks, sizeof(struct s3_buffer_pool_block));
```

Comment:
this is nice since IIRC this will put all blocks on the same allocation.... Now just need to make sure this stays on the same numa node at some point.

Reply:
I pushed the allocations to prepare stage, so it will run from io prep threads. 
so in practice it can allocate from different numa nodes for bigger blocks, but then same block can be reused by different nodes. 
In future we can tie blocks to specific numa nodes, something as trivial as keeping a list of blocks per numa node will work